### PR TITLE
Fix #9184: Support fmtIntSizeDec/fmtIntSizeBin FormatOptions

### LINKED ISSUE #9184: std.fmt.fmtIntSizeDec() and fmtIntSizeBin() do not consider FormatOptions
Ultimately `formatSizeImpl()` is used to implement `fmtIntSizeDec()` or `fmtIntSizeBin()` and it does not consider FormatOptions and probably should.

motivating example (from @espindola  via IRC):
```zig
const std = @import("std");

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();

    var buf: [32]u8 = undefined;
    var slice = buf[0..];
    const partial = try std.fmt.bufPrint(slice, "{}", .{std.fmt.fmtIntSizeDec(42 * 1000)});

    try stdout.print("X {s:<14} Y\n", .{partial});
    try stdout.print("X {:<14} Y\n", .{std.fmt.fmtIntSizeDec(42 * 1000)});
}
```

```sh
$ zig run foo.zig
X 42kB           Y
X 42kB Y
```

@andrewrk you mentioned in IRC:
```
it should also be renamed to remove the redundant fmt
```

There is a pattern in `std.fmt` where all `pub fn fmt*()` functions return a `Formatter()` object; a grep follows. Do you want these  renamed like `fmtFirstSecond()` → `firstSecond()`?
```
pub fn fmtSliceHexLower(bytes: []const u8) std.fmt.Formatter(formatSliceHexLower) {
pub fn fmtSliceHexUpper(bytes: []const u8) std.fmt.Formatter(formatSliceHexUpper) {
pub fn fmtSliceEscapeLower(bytes: []const u8) std.fmt.Formatter(formatSliceEscapeLower) {
pub fn fmtSliceEscapeUpper(bytes: []const u8) std.fmt.Formatter(formatSliceEscapeUpper) {
pub fn fmtIntSizeDec(value: u64) std.fmt.Formatter(formatSizeDec) {
pub fn fmtIntSizeBin(value: u64) std.fmt.Formatter(formatSizeBin) {
pub fn fmtDuration(ns: u64) Formatter(formatDuration) {
```

### PR DESCRIPTION
**Pull request by yujiri8**
I'm new to this. Am I going about it the right way, not changing the API of fmtIntSizeDec and fmtIntSizeBin themselves? I'm confused about the API, is this how it's meant to be used?
```zig
const std = @import("std");
const fmt = std.fmt;
const debug = std.debug;
const io = std.io;

pub fn main() !void {
    var stdout = io.getStdOut().writer();
    try fmt.fmtIntSizeBin(10000).format("", fmt.FormatOptions{.width = 12, .alignment = .Center}, stdout);
    // debug.print("{s}", .{result});
}
```

(It currently prints `9.765625    KiB     `. I assume the intended behavior would be to put no space inbetween the number and `KiB` and center the concanetation?)

Edit: nevermind my question about the API. I think I figured out how it's meant to be used, I didn't realize the format placeholders had that much flexibility

### CODE REVIEW

File: lib/std/fmt.zig [hunk 1]
```diff
@@ -876,8 +876,11 @@ fn formatSizeImpl(comptime radix: comptime_int) type {
             writer: anytype,
         ) !void {
             _ = fmt;
+            // The worst case in terms of space needed is... TODO.
+            var buf: [1000]u8 = undefined;
             if (value == 0) {
-                return writer.writeAll("0B");
+                mem.copy(u8, buf[0..], "0B");
+                return formatBuf(buf[0..2], options, writer);
```
Comment (Vexu) [inline lib/std/fmt.zig:883]:
```suggestion
                return formatBuf("0B", options, writer);
```

File: lib/std/fmt.zig [hunk 2]
```diff
@@ -876,8 +876,11 @@ fn formatSizeImpl(comptime radix: comptime_int) type {
             writer: anytype,
         ) !void {
             _ = fmt;
+            // The worst case in terms of space needed is... TODO.
+            var buf: [1000]u8 = undefined;
```
Comment (Vexu) [inline lib/std/fmt.zig:880]: The worst case for `formatFloatDecimal` seems to be 32 bytes + 3 bytes for the suffix.

**Review by Vexu — CHANGES_REQUESTED** [review]
The float value and the plain `B` suffix are still being written directly to the writer. Instead everything should be written into a `std.io.fixedBufferStream` backed by the buffer on stack and then the `fbs.getWritten()` should be written into the original writer using `formatBuf`.

**Comment by yujiri8** [comment]
Using fixedBufferStream now. This almost works, but the tests are failing with:
```
./lib/std/fmt.zig:1662:13: error: expected type 'std.io.NullWriter', found '@typeInfo(@typeInfo(@TypeOf(std.fmt.formatSizeImpl(1000).f)).Fn.return_type.?).ErrorUnion.error_set'
            try format_fn(self.data, fmt, options, writer);
            ^
./lib/std/fmt.zig:461:32: note: called from here
        return try value.format(actual_fmt, options, writer);
                               ^
./lib/std/fmt.zig:181:23: note: called from here
        try formatType(
                      ^
./lib/std/fmt.zig:1893:11: note: called from here
    format(counting_writer.writer(), fmt, args) catch |err| switch (err) {};
          ^
./lib/std/fmt.zig:1900:40: note: called from here
    const size = math.cast(usize, count(fmt, args)) orelse return error.OutOfMemory;
                                       ^
./lib/std/testing.zig:201:42: note: called from here
    const result = try std.fmt.allocPrint(allocator, template, args);
                                         ^
./test.zig:10:26: note: called from here
    try testing.expectFmt("", "{d}", .{fmt.fmtIntSizeDec(42)});
                         ^
./test.zig:7:21: note: called from here
pub fn main() !void {
                    ^
./lib/std/io/fixed_buffer_stream.zig:16:38: note: 'error.NoSpaceLeft' not a member of destination error set
        pub const WriteError = error{NoSpaceLeft};
                                     ^
```
Replacing every usage of `try` in the function with `catch unreachable` solves it, but I don't think that's correct since it was using try before when calling formatFloatDecimal. Meanwhile my own test case like this works fine:
```zig
const std = @import("std");
const fmt = std.fmt;
const debug = std.debug;
const io = std.io;
const testing = std.testing;

pub fn main() !void {
    debug.print("{s: ^20}", .{fmt.fmtIntSizeDec(18446744073709551615)});
    // try testing.expectFmt("", "{d}", .{fmt.fmtIntSizeDec(42)});
}
```
It seems to be only testing.expectFmt that doesn't like it.

**Comment by yujiri8** [comment]
Updated! The tests pass now and testing.expectFmt seems to work. Thanks for your help!

**Review by Vexu — APPROVED** [review]
Could you add  a new test for alignment and width like was done in #10361?

**Comment by yujiri8** [comment]
Done! Four new test cases.

**Pull request merged**